Fixed deepcopy for Atom.

ywu opened the pull request:
This PR is to fix a deepcopy-related bug that results in error for calls like `deepcopy(atom)` if Atom.__deepcopy__ is not defined.

swails commented:
Are you sure this is needed? I had thought that deepcopy used `__copy__` if the `__deepcopy__` method wasn't defined. 

Can you share some code that this fixes?

ywu commented:
I added a unittest to hopefully demonstrate the difference (and fixed a bug in my previous change :blush:): Running the unittest with and without the `Atom.__deepcopy__` method defined, you should see `deepcopy(atom)` doesn't call `Atom._copy` if `Atom.__deepcopy__` is not defined.

swails commented:
I had operated under the assumption that `copy.deepcopy()` would just use `__copy__` if it `__deepcopy__` wasn't available, so if this is a problem for `Atom`, then it's a problem for every other class, too.

Can you give me a real-world script where the `copy.deepcopy()` semantics produces incorrect behavior?
